remote-wallet/src/remote_wallet.rs

```diff
@@ -322,20 +324,61 @@ impl RemoteWalletInfo {
     }
 }
 
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct DerivationPathComponent(u32);
+
+impl DerivationPathComponent {
+    pub const HARDENED_BIT: u32 = 1 << 31;
+
+    pub fn as_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+impl From<u32> for DerivationPathComponent {
+    fn from(n: u32) -> Self {
+        Self(n | Self::HARDENED_BIT)
+    }
+}
+
+impl FromStr for DerivationPathComponent {
+    type Err = RemoteWalletError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Replace str::splitn() with str::strip_suffix() once stabilized
+        let parts: Vec<_> = s.splitn(2, '\'').collect();
+        if parts.len() == 2 {
+            eprintln!("all path components are promoted to hardened representation");
+        }
+        parts[0].parse::<u32>().map(|ki| ki.into()).map_err(|_| {
+            RemoteWalletError::InvalidDerivationPath(format!(
+                "failed to parse path component: {:?}",
+                s
+            ))
+        })
+    }
+}
+
+impl std::fmt::Display for DerivationPathComponent {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(fmt, "{}", self.0)
+    }
+}
```

Comment:
Can we rework this? Right now, if I debug print the DerivationPath after parsing, it looks like this for `?key=1`, which is both confusing and actually incorrect (due to the apostrophe) 😞 :
```
m/44'/501'/DerivationPathComponent(2147483649)'
```

Would prefer this output for at least Display, and maybe Debug as well:
```
m/44'/501'/1'
```

You'll probably want to move the addition of the apostrophes into Display/Debug for DerivationPathComponent, currently here:
https://github.com/solana-labs/solana/blob/d42247c6524f249247ab39eb28cfe0da3069362e/remote-wallet/src/remote_wallet.rs#L333

Reply:
Updated! https://github.com/solana-labs/solana/pull/10725/commits/493c19a5fbfa6b12f260c9d1a1eef37514927f78#diff-7ef79c2848cdef4bdf705e9d9000c5dfR660-R662